### ISSUE TITLE
Inline backend not working on master

### ISSUE THREAD

**Issue by michaelaye** [matplotlib/matplotlib#4774]
I just installed master (who could resist those spanking new colormaps?? ;) ) and have trouble with the inline backend. `nbagg` works fine, but when activating `%matplotlib inline`  in the notebook, the MPL object is being returned and REPL'ed, but no graph/image is shown.
I quickly pulled it by doing 
```bash
pip uninstall matplotlib
Uninstalling matplotlib-1.5.dev1:
  /Users/klay6683/miniconda3/lib/python3.4/site-packages/matplotlib-1.5.dev1-py3.4-macosx-10.5-x86_64.egg
Proceed (y/n)? y
  Successfully uninstalled matplotlib-1.5.dev1
conda install matplotlib
```
and tried the same again, and it works, so it's apparently not the notebook's fault.

Here's the mini code I ran:
```python
%matplotlib inline
img = np.arange(10000).reshape(100,100)
plt.figure()
plt.imshow(img, cmap='gray')
```

In other words, MPL master reacts exactly as it would look if I comment out the %matplotlib magic above when running with MPL 1.4.3.

**Comment by tacaswell** [comment]
If you throw in a `plt.show` in works again (I also just saw this testing a strange pandas/IPython bug).

I suspect the issue is that due to the changes in how the interactive code works we no longer have an implicit `plt.show` because we pulled `draw_if_interactive` out of pyplot.  Right now that is taken care in each of the backend's figure manager creation which is not great.  This could/should probably be centralized into `plt.figure`.

attn @takluyver 

This is a some what major re-work on our side and I think 3 lines on your side.

**Comment by michaelaye** [comment]
is there some more info about pulling `draw_if_interactive` out of pyplot? I always thought that pyplot was the programmatically better alternative to `pylab`, and therefore it should support interactive stuff, shouldn't it? (Maybe too naive view?). You certainly are not advocating to go back to using pylab mode now, are you?

**Comment by tacaswell** [comment]
Sorry, I can see how that was a confusing statement. 

It has been replaced by something much more powerful which gets the auto-redraw feature on _any_ artist change, not just the sub-set of commands that are wrapped by `pyplot`.  In short, `Artists` now know if they have changed since the last time they were drawn and `plt.ion()` does it's best to install hooks to trigger redraws on state change.

See https://github.com/matplotlib/matplotlib/blob/master/doc/users/whats_new/2015-05_interactive_OO.rst

**Comment by michaelaye** [comment]
Thanks. Maybe silly question but, if the `Artists` already know their state has changed, why not let THEM trigger the redraw instead of implementing something in between that checks if a state has changed?

**Comment by WeatherGod** [comment]
Because we don't want to trigger redraws if the library is not being used
in an interactive environment such as scripting, or if the user did not
turn interactivity on. Also, some environments handle redraws themselves
such as IPython. This is a proper separation of responsibilities. Artists
simply say "I'm stale", and other things decide what to do about that.

**Comment by michaelaye** [comment]
ah, cool. thanks for explaining!

**Comment by tacaswell** [comment]
Because managing the cadence of the redraws should be the responsibility of the embedding application (the most common of which is `pyplot`).  You also get fun bugs if they canvas you are using does not implement an asynchronous `draw_idle`.

There are also use cases coming where it will make sense to have figures which do not have a canvas (thinking about the export driven uses cases), where the binding to the canvas may happen very late, or where figure may know about mulitple canvas (thinking of the IPython kernel case where it would be nice to have the figure able to shown as an inline png, as a nbagg widget, or as a full Qt window depending on which front end has connected to the kernel).

**Comment by takluyver** [comment]
Sorry, I hadn't got round to looking at this before. My job at Berkeley has officially finished, so my activity on IPython will be a bit sporadic for a few weeks until my new job in Southampton begins (planned for 1st October).

If there's something we should still be doing in IPython, please do open an issue so we don't forget it. Thanks!

**Comment by tacaswell** [comment]
@takluyver I dealt with it completely on our side in a way that I think makes a decent amount of sense.

**Comment by takluyver** [comment]
Awesome, thanks!
